homeassistant/components/modem_callerid/config_flow.py

```diff
@@ -0,0 +1,142 @@
+            for port in ports:
+                if port.device == user_input[CONF_DEVICE]:
+                    entry, errors = await self.validate_input(
+                        user_input=user_input,
+                        dev_path=port.device,
+                        unique_id=_generate_unique_id(port),
+                    )
+                    if errors is None:
+                        return entry
+        existing_devices = [
+            entry.data[CONF_DEVICE] for entry in self._async_current_entries()
+        ]
+        unused_ports = [
+            usb.human_readable_device_name(
+                port.device,
+                port.serial_number,
+                port.manufacturer,
+                port.description,
+                port.vid,
+                port.pid,
+            )
+            for port in ports
+            if port.device not in existing_devices
+        ]
+        if not unused_ports:
+            return self.async_abort(reason="no_devices_found")
+
+        if user_input is not None:
+            port = ports[unused_ports.index(str(user_input.get(CONF_DEVICE)))]
+            dev_path = await self.hass.async_add_executor_job(
+                usb.get_serial_by_id, port.device
+            )
+            entry, errors = await self.validate_input(
+                dev_path=dev_path, unique_id=_generate_unique_id(port)
+            )
+            if errors is None:
+                return entry
+        user_input = user_input or {}
+        schema = vol.Schema({vol.Required(CONF_DEVICE): vol.In(unused_ports)})
+        return self.async_show_form(step_id="user", data_schema=schema, errors=errors)
+
+    async def async_step_import(self, config: dict[str, Any]) -> FlowResult:
+        """Import a config entry from configuration.yaml."""
+        if self._async_current_entries():
+            _LOGGER.warning(
+                "Loading Modem_callerid via platform setup is deprecated; Please remove it from your configuration"
+            )
+        if CONF_DEVICE not in config:
+            config[CONF_DEVICE] = DEFAULT_PORT
+
+        return await self.async_step_user(config)
+
+    async def validate_input(self, user_input=None, dev_path=None, unique_id=None):
+        """Handle common flow input validation."""
+        user_input = user_input or {}
+        errors = {}
+        self._async_abort_entries_match({CONF_DEVICE: dev_path})
+        await self.async_set_unique_id(unique_id)
+        self._abort_if_unique_id_configured(updates={CONF_DEVICE: dev_path})
+        try:
+            api = PhoneModem()
+            await api.test(dev_path)
+            entry = self.async_create_entry(
+                title=user_input.get(CONF_NAME, DEFAULT_NAME),
+                data={CONF_DEVICE: dev_path},
+            )
+            return entry, None
+        except EXCEPTIONS:
+            errors["base"] = "cannot_connect"
+            return None, errors
```

Comment:
```suggestion
        except EXCEPTIONS:
            errors["base"] = "cannot_connect"
            return None, errors
        else:
            entry = self.async_create_entry(
                title=user_input.get(CONF_NAME, DEFAULT_NAME),
                data={CONF_DEVICE: dev_path},
            )
            return entry, None        
```

Actually is kinda unexpected to create an entry from a validator.

It would be better to return the validated data instead and have the caller create the entry.